### ISSUE TITLE
Historical records + meta tags VS push system

### ISSUE THREAD

**Issue by comboy** [olalonde/proof-of-solvency#1]
Hi,
Thanks for your great work.
I would suggest to put verification data directly into the page source and not using meta tags to provide links where it can be fetched (script tag perhaps?). This way exchange wouldn't know who and when is checking the proof which should improve security in some extreme cases.

**Comment by olalonde** [comment]
Good point! One drawback I see is that it would make it slightly more difficult for exchanges to implement and might increase page load times for large proofs (URLs can be cached). The best compromise I can think of would be to use an HTML tag that automatically fetches an URL (like `<script>`/`<style>` do and make that fetched data available through Javascript so that an extension can automatically verify the proof). Perhaps `<script src="/proof.json">` would be a good choice.

**Comment by comboy** [comment]
<tt><script src=""></tt> seems nice. I don't think size will be a problem since these should only grow logarithmically, and I'm not sure url cache is worth the additional request for data of this size (especially with https), but I admit it seems more elegant like this (in a separate file). 

Anyway, that was just a random thought, thanks again for doing this.

**Comment by jaekwon** [comment]
My exchange doesn't render any content server-side, and my *-assets.proof may grow pretty large.

> > This way exchange wouldn't know who and when is checking the proof which should improve security in some extreme cases.

I don't understand. How does putting verification data directly in the page source help in any way?

The server _must_ know who is checking the liabilities partial-tree.
External robots can scrape the /*-root.json file regularly to guard against the server dynamically generating the root. The browser plugin should communicate with this external service somehow & verify the root. Anything less & you're putting too much trust into the exchange, since it's serving all the scripts.

Also, please don't use `<script>` tags.

**Comment by olalonde** [comment]
> My exchange doesn't render any content server-side, and my *-assets.proof may grow pretty large.

I agree that putting too much data directly in the source would probably be a bad idea. With separate files, you can at least leverage the browser cache.

> The server must know who is checking the liabilities partial-tree.

That's true but I think the reasoning behind this is that if you force users to automatically download their tree, there is no way for you as an operator to tell who is really checking and who isn't since everyone downloads the file anyways. You can't single out users who are not checking their tree and "remove" them from the proof (of course, you can still single out users who haven't logged in in a long time but it makes it more difficult to cheat overall).

I don't think the proofs would necessarily need to be directly in the source but it would be nice to use a tag that makes web browsers automatically fetch the URL. `<script src="/account/partial-tree.json"></script>` for example (even `<img>` would work although it would not be very elegant!). 

Maybe another tag would be better suited for this or maybe forcing users to download their tree isn't that important after all.

> External robots can scrape the /*-root.json file regularly to guard against the server dynamically generating the root. The browser plugin should communicate with this external service somehow & verify the root. Anything less & you're putting too much trust into the exchange, since it's serving all the scripts.

Yes I think this would really be needed. I've written an issue about this in the extension repository if I recall correctly.

PS: @jaekwon really glad you are working on implementing the scheme... your comments are really important. Let me know if my explanation makes sense.
